modules/encrypt/classes/Kohana/Encrypt.php

```diff
@@ -10,91 +9,93 @@
+	 * Encrypt class instances
+     * @var array
      */
     public static $instances = [];
 
+	/**
+	 * Engine Name
+	 * @var string
+	 */
+    private $_name;
+
     /**
-     * @var  Kohana_Encrypt_Engine Encryption engine
+	 * Encryption engine
+     * @var Kohana_Encrypt_Engine
      */
-    private $_engine = NULL;
+    private $_engine;
 
     /**
      * Returns a singleton instance of Encrypt. An encryption key must be
      * provided in your "encrypt" configuration file.
      *
      *     $encrypt = Encrypt::instance();
      *
-     * @param string $name configuration group name
-     * @param array|null $config
-     * @return Encrypt
+     * @param  string 		$name 		Configuration group name
+     * @param  array|null 	$config		Configuration
+     * @return self
      * @throws Kohana_Exception
      */
-    public static function instance(string $name = NULL, array $config = NULL)
+    public static function instance(string $name = NULL, array $config = NULL) : self
     {
         if ($name === NULL)
         {
             // Use the default instance name
-            $name = Encrypt::$default;
+            $name = self::$default;
         }
 
-        if (!isset(Encrypt::$instances[$name]))
+        if (!isset(self::$instances[$name]))
         {
             if ($config === NULL)
             {
                 // Load the configuration data
                 $config = Kohana::$config->load('encrypt')->$name;
             }
 
-            if (!isset($config[Kohana_Encrypt_Engine::CONFIG_KEY]))
+            if (!isset($config['key']))
             {
                 // No default encryption key is provided!
                 throw new Kohana_Exception('No encryption key is defined in the encryption configuration group: :group',
                     [':group' => $name]);
             }
 
             // Create a new instance
-            Encrypt::$instances[$name] = new self($name, $config);
+			self::$instances[$name] = new self($name, $config);
         }
 
-        return Encrypt::$instances[$name];
+        return self::$instances[$name];
     }
 
     /**
      * Creates a new Encrypt Engine instance.
      *
-     * @param array $config
-     * @param string $name
+	 * @param string $name		Engine Name
+     * @param array  $config	Configuration
      */
     private function __construct(string $name, array $config)
     {
-        if (!isset($config[Kohana_Encrypt_Engine::CONFIG_TYPE]))
-        {
-            $config[Kohana_Encrypt_Engine::CONFIG_TYPE] = Encrypt_Engine_Openssl::TYPE;
-        }
-
-        $this->_name = $name;
+    	// Get Driver Type
+		$config['type'] = $config['type'] ?? self::$default_engine;
 
         // Set the engine class name
-        $engine_name = 'Encrypt_Engine_' . ucfirst($config[Kohana_Encrypt_Engine::CONFIG_TYPE]);
+        $engine_name = 'Encrypt_Engine_' . ucfirst($config['type']);
 
         // Create the engine class
-        $this->setEngine(new $engine_name($config));
```

Comment:
Because function setEngine only accepts Kohana_Encrypt_Engine, this was validation that passed class implements Kohana_Encrypt_Engine, which you removed, I don't know why. Can you revert this?

Reply:
You didn't validate it you just required it in the function parameters. This would throw a php error if engine is not an instance - which is not good. So basicly your function was just a setter which didn't perform any actions. According to [google](https://web.archive.org/web/20140625191431/https://developers.google.com/speed/articles/optimizing-php) this reduces speed and is considered a bad practice

But i agree we need to validate... i will add a check if the class is availiable/exists AND after calling if is an instance of Kohana_Encrypt_Engine. If not it will throw a Kohana Exception stating that the engine is not valid.